sdk/go/pulumi/types.go

```diff
@@ -247,11 +256,13 @@ func (o *OutputState) await(ctx context.Context) (interface{}, bool, bool, []Res
 		//
 		// NOTE: this isn't exactly type safe! The element type of the inner output really needs to be assignable to
 		// the element type of the outer output. We should reconsider this.
-		ov, ok := o.value.(Output)
-		if !ok {
+		if ov, ok := o.value.(Output); ok {
+			deps := o.deps
+			o = ov.getState()
+			o.deps = mergeDependencies(o.deps, deps)
+		} else {
 			return o.value, true, o.secret, o.deps, nil
 		}
```

Comment:
Nit: I think this style is easier to read. Feel free to ignore. 

```suggestion
		ov, ok := o.value.(Output)
		if !ok {
			return o.value, true, o.secret, o.deps, nil
		}
		deps := o.deps
		o = ov.getState()
		o.deps = mergeDependencies(o.deps, deps)
```